File: core/SimAgent.java
/**
 * Copyright (c) 2020 Team 3260
 */

import java.io.*;
import java.net.*;

/**
 * Comms agent responsible for communicating with the sim. Continuously sends commands to the sim to update the vehicle
 * and continuously listens for vehicle state back from the sim.
 */
public class SimAgent
{
    public SimCommands commands;
    public SimState state;
    private UdpNode comms;

    public SimAgent(int rxPort, String txIp, int txPort)
    {
        state = new SimState();
        commands = new SimCommands();
        try
        {
            comms = new UdpNode(rxPort, txIp, txPort);
        }
        catch (Exception e)
        {
            System.out.println("SimAgent: Failed to init UDP node");
        }
    }

    /**
     * Sends the sim's current state stored in `state`
     */
    public void txSimCommands()
    {
        // Translate commands to JSON
        String msg = commands.toJson();

        // Send string
        try
        {
            comms.send(msg);
            // System.out.println("SimAgent: Transmit commands " + msg);
        }
        catch (IOException e)
        {
            System.out.println("SimAgent: Failed to tx");
        }
    }

    /**
     * Receives the sim's commands and stores it into `commands`
     */
    public void rxSimState()
    {
        try
        {
            // Receive commands
            String msg = comms.receive();

            // Parse received command from JSON to state
            state.fromJson(msg);
            // System.out.println("SimAgent: Received " + msg);
        }
        catch (IOException e)
        {
            // System.out.println("SimAgent: Failed to rx");
        }
    }
}

File: sim/src/SimAgent.cpp
/**
 * Copyright (c) 2020 Team 3260
 */

#include <string>
#include "SimAgent.h"

/** Connection timeout in seconds */
#define CONNECTED_TIMEOUT 0.5


SimAgent::SimAgent(const ConfigReader& config) :
        AbstractAgent(config.simView.port, config.sim.comms.ip, config.sim.comms.simViewPort),
        _simState{},
        _verbose(config.verbose)

{
    std::cout << "Rx from sim at " << config.simView.ip << ":" << config.simView.port << std::endl;
    std::cout << "Tx to sim at " << config.sim.comms.ip << ":" << config.sim.comms.simViewPort << std::endl;
}



void SimAgent::txHeartbeat()
{
    // Transmit state
    std::string msg = "{}";

    if (_verbose)
    {
        printf("SimAgent: Sent %s\n", msg.c_str());
    }

    _comms->send(msg);
}



SimState SimAgent::rxSimState()
{
    std::string msg = _comms->receive();
    if (msg.length() > 0 && msg[0] == '{')
    {
        _connected = true;

        if (_verbose)
        {
            printf("SimAgent: Received command %s\n", msg.c_str());
        }

        // Translate received commands from JSON and store into _simState
        _simState.fromJson(msg);

        // Save time of last rx
        _prevRxTime = getCurrentTime();
    }
    else
    {
        if (getCurrentTime() - _prevRxTime > CONNECTED_TIMEOUT)
        {
            _connected = false;
            _simState.clear();
        }
    }

    return _simState;
}
